src/System.Memory/src/System/Buffers/ReadOnlyBuffer_helpers.cs

```diff
@@ -0,0 +1,305 @@
+                    return default;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal SequencePosition Seek(SequencePosition start, SequencePosition end, long count, bool checkEndReachable = true)
+        {
+            var startIndex = start.Index;
+            var endIndex = end.Index;
+            var type = GetBufferType();
+
+            startIndex = GetIndex(startIndex);
+            endIndex = GetIndex(endIndex);
+
+            switch (type)
+            {
+                case BufferType.MemoryList:
+                    if (start.Segment == end.Segment && endIndex - startIndex >= count)
+                    {
+                        // end.Index >= bytes + Index and end.Index is int
+                        return new SequencePosition(start.Segment, startIndex + (int)count);
+                    }
+                    return SeekMultiSegment((IMemoryList<byte>) start.Segment, startIndex, (IMemoryList<byte>) end.Segment, endIndex, count, checkEndReachable);
+
+
+                case BufferType.OwnedMemory:
+                case BufferType.Array:
+                    if (endIndex - startIndex >= count)
+                    {
+                        // end.Index >= bytes + Index and end.Index is int
+                        return new SequencePosition(start.Segment, startIndex + (int)count);
+                    }
+
+                    ThrowHelper.ThrowArgumentOutOfRangeException_CountOutOfRange();
+                    return default;
+
+                default:
+                    ThrowHelper.ThrowInvalidOperationException_UnexpectedSegmentType();
+                    return default;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.NoInlining)]
+        private static SequencePosition SeekMultiSegment(IMemoryList<byte> start, int startIndex, IMemoryList<byte> end, int endPosition, long count, bool checkEndReachable)
+        {
+            SequencePosition result = default;
+            var foundResult = false;
+            var current = start;
+            var currentIndex = startIndex;
+
+            while (current != null)
+            {
+                // We need to loop up until the end to make sure start and end are connected
+                // if end is not trusted
+                if (!foundResult)
+                {
+                    var memory = current.Memory;
+                    var currentEnd = current == end ? endPosition : memory.Length;
+
+                    memory = memory.Slice(0, currentEnd - currentIndex);
+                    // We would prefer to put position in the beginning of next segment
+                    // then past the end of previous one, but only if we are not leaving current buffer
+                    if (memory.Length > count ||
+                       (memory.Length == count && current == end))
+                    {
+                        result = new SequencePosition(current, currentIndex + (int)count);
```

Comment:
Could this overflow?

Reply:
`count <= currentEnd - currentIndex` -> `currentIndex + count <= currentIndex  + currentEnd - currentIndex` -> `currentIndex + count <= currentEnd`

Reply:
How do we know `count <= currentEnd - currentIndex` here? It is passed in.

Reply:
```
memory = memory.Slice(0, currentEnd - currentIndex);
if (memory.Length > count || (memory.Length == count && current == end))
```


`memory.Length == currentEnd - currentIndex && memory.Length > count || (memory.Length == count `

Reply:
Ah yes. Got it. Thanks for the explanation.